### PR TITLE
onCellsChanged is triggered twice

### PR DESCRIPTION
**Pull request by qiufeihong2018**
…red in a single text cell and onCellsChanged is triggered, changes become an array of one element, but a two-element array #289

resolved

image

### CODE REVIEW

File: src/lib/Functions/focusLocation.ts [hunk 1]
```diff
@@ -4,11 +4,21 @@ import { tryAppendChange } from './tryAppendChange';
...
 
-export function focusLocation(state: State, location: Location, applyResetSelection = true): State {
-    if (state.focusedLocation && state.currentlyEditedCell) {
-        state = tryAppendChange(state, state.focusedLocation, state.currentlyEditedCell);
+export function focusLocation(state: State, location: Location, applyResetSelection = true, keyCode?: keyCodes): State {
+    // Need to determine if there is an edited cell &&The cell gets focus &&did not press Enter
+    if (
+      state.focusedLocation &&
+      state.currentlyEditedCell &&
+      !(keyCode === keyCodes.ENTER)
```
Comment (Michota) [inline src/lib/Functions/focusLocation.ts]: More clear me.
```suggestion
      keyCode !== keyCodes.ENTER
```

**Review by Michota — CHANGES_REQUESTED** [review]
I would change the way it checks the `Enter` keycode. Instead of checking for a keycode, i would suggest to pass boolean, that is based on if `Enter` was clicked, something like `isEnter = keyCode === keyCodes.Enter`, which would be more clear in my opinion, but its not necessary.

**Review by Michota — APPROVED** [review]

**Comment by mitchell-robinson** [comment]
@Michota I'm not sure this is fixed properly, I still receive two change events when hitting `Enter`

In the video below, the first cell I edit, I hit `Enter` -- notice the array with two events where I'm logging the `CellChange[]` array passed to the `onCellsChanged` handler.

When I edit the second cell, I blur to another cell -- notice the single event logged.

https://github.com/silevis/reactgrid/assets/101780654/1680262b-d866-4583-8602-ec0fb38c20b7

Running latest 4.1.3

**Comment by qiufeihong2018** [comment]
@mitchell-robinson 
Surprisingly, I did see the correct result in the example. When you updated to 4.1.3, this problem still occurred. Could you please repeat it in src/test/TestGrid.tsx? Or take a look at the faulty code in your project in CodePen?

**Comment by mitchell-robinson** [comment]
@qiufeihong2018 @Michota
I was unable to reproduce the error in `src/test/TestGrid.tsx`.

However, I created two fresh projects, and used the [Getting Started example](https://reactgrid.com/docs/4.0/1-getting-started/) from the docs to set up a quick reproducible example:

One using Create React App + Typescript:
https://github.com/mitchell-robinson/cra-ts-reactgrid-test

The other using NextJs:
https://github.com/mitchell-robinson/nextjs-reactgrid-test

Both just setup and run the example from the docs, and console log the `changes` from the onCellsChanged handler. I'm seeing 2 events when hitting enter here in each scenario. Are either of you able to pull those repos, run, and reproduce?

thanks in advance

*Update: Went ahead and made this as a codepen from the docs example as well:
https://codesandbox.io/p/sandbox/reactgrid-duplicate-change-on-enter-gvpwl5

Forked from the docs codepen

**Comment by kateEcobb** [comment]
This fix hasn't been released yet, so it shouldn't be marked as resolved. Is there a plan to release this soon?

**Comment by Michota** [comment]
@kateEcobb im working on v5 now.
After releasing v5-alpha i will come back to v4, and will fix this issue.

**Pull request merged**